ast.ts

```diff
@@ -42,6 +53,8 @@ export type Expr<A> =
   | {  a?: A, tag: "index", obj: Expr<A>, index: Expr<A> }
   | {  a?: A, tag: "method-call", obj: Expr<A>, method: string, arguments: Array<Expr<A>> }
   | {  a?: A, tag: "construct", name: string }
+  | {  a?: A, tag: "comprehension", type: Type, lhs: Expr<A>, item: string, iterable: Expr<A>, ifcond?: Expr<A> } // comprehension expression
```

Comment:
What's this `Type`? Can you have an `int` comprehension?

Reply:
`type` here is the type of the comprehension (i.e. what type of object the comprehension will return). E.g. `[a for a in [1, 2, 3]]` has type `list` (assuming `list` is a valid type implemented by the list group), while `(a for a in [1, 2, 3])` has type `generator`.